File: Core/src/main/java/jokatu/game/event/dialog/DialogRequest.java
package jokatu.game.event.dialog;

import jokatu.game.event.GameEvent;
import jokatu.game.input.Input;
import jokatu.game.player.Player;
import jokatu.ui.Dialog;
import jokatu.ui.Form;
import ophelia.function.ExceptionalBiConsumer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class DialogRequest<P extends Player, I extends Input> implements GameEvent {

	private final Class<I> expectedInput;
	private final P player;
	private final String title;
	private final String message;
	private final Form form;
	private final ExceptionalBiConsumer<? super I, ? super P, ?> consumer;
	private final boolean cancellable;

	DialogRequest(
			@NotNull Class<I> expectedInput,
			@NotNull P player,
			@NotNull String title,
			@NotNull String message,
			@NotNull ExceptionalBiConsumer<? super I, ? super P, ?> consumer,
			@Nullable Form form,
			boolean cancellable
	) {
		this.expectedInput = expectedInput;
		this.player = player;
		this.title = title;
		this.message = message;
		this.form = form;
		this.consumer = consumer;
		this.cancellable = cancellable;
	}

	void accept(@NotNull I i) throws Exception {
		consumer.accept(i, player);
	}

	@NotNull
	Class<I> getInputClass() {
		return expectedInput;
	}

	@NotNull
	Dialog getDialog() {
		return new Dialog(title, message, form, cancellable);
	}

	@NotNull
	Player getPlayer() {
		return player;
	}
}

File: Core/src/main/kotlin/jokatu/game/event/dialog/DialogRequest.kt
package jokatu.game.event.dialog

import jokatu.game.event.GameEvent
import jokatu.game.input.Input
import jokatu.game.player.Player
import jokatu.ui.Dialog
import jokatu.ui.Form
import ophelia.function.ExceptionalBiConsumer
import kotlin.reflect.KClass

class DialogRequest<P : Player, I : Input> internal constructor(
		val inputClass: KClass<I>,
		val player: P,
		private val title: String,
		private val message: String,
		private val consumer: ExceptionalBiConsumer<in I, in P, *>,
		private val form: Form?,
		private val cancellable: Boolean
) : GameEvent {

	val dialog: Dialog
		get() = Dialog(title, message, form, cancellable)

	@Throws(Exception::class)
	fun accept(i: I) {
		consumer.accept(i, player)
	}

	internal fun getPlayer(): Player {
		return player
	}
}
